### PR TITLE
feat: [STO-4898]: Gate STO access by license

### PR DESCRIPTION
**Pull request by ahokinson**
### Summary

The STO Module needs to have access gated by license and no longer by feature flags. This request also removes those feature flags.

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 9 Issues 
* Code issue 9 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/e864badd0b95ab81c060b15b85f15701/pull-request/11635)

**Comment by ahokinson** [comment]
retrigger titlecheck

File: src/modules/70-pipeline/components/ExecutionCard/ExecutionCard.tsx [hunk 1]
```diff
@@ -154,7 +152,7 @@ export default function ExecutionCard(props: ExecutionCardProps): React.ReactEle
     useParams<PipelineType<PipelinePathProps>>()
   const history = useHistory()
   const { getString } = useStrings()
-  const SECURITY = useFeatureFlag(FeatureFlag.SECURITY)
+  const SECURITY = true
```
Comment (atmollo-harness) [inline src/modules/70-pipeline/components/ExecutionCard/ExecutionCard.tsx]: If this variable is being defaulted to true you can probably just remove it

File: src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx [hunk 2]
```diff
@@ -52,8 +52,8 @@ export default function ExecutionTabs(props: ExecutionTabsProps): React.ReactEle
   const { view } = useQueryParams<ExecutionQueryParams>()
   const { updateQueryParams } = useUpdateQueryParams<ExecutionQueryParams>()
   const opaBasedGovernanceEnabled = useFeatureFlag(FeatureFlag.OPA_PIPELINE_GOVERNANCE)
-  const stoCDPipelineSecurityEnabled = useFeatureFlag(FeatureFlag.STO_CD_PIPELINE_SECURITY)
-  const stoCIPipelineSecurityEnabled = useFeatureFlag(FeatureFlag.STO_CI_PIPELINE_SECURITY)
+  const stoCDPipelineSecurityEnabled = true
+  const stoCIPipelineSecurityEnabled = true
```
Comment (atmollo-harness) [inline src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx]: same things for these two

File: src/modules/85-sto-steps/RouteDestinations.tsx [hunk 3]
```diff
@@ -53,6 +65,111 @@ const moduleParams: ModulePathParams = {
...
 
+// License
+
+featureFactory.registerFeaturesByModule('sto', {
+  features: [
+    FeatureIdentifier.MAX_TOTAL_BUILDS,
+    FeatureIdentifier.MAX_BUILDS_PER_MONTH,
+    FeatureIdentifier.ACTIVE_COMMITTERS
+  ],
+  renderMessage: (props, getString, additionalLicenseProps = {}) => {
```
Comment (atmollo-harness) [inline src/modules/85-sto-steps/RouteDestinations.tsx:76]: It would be cleaner to move this function to a separate file rather than adding it to the RouteDestinations for the sake of readability

File: src/framework/featureStore/FeatureIdentifier.ts [hunk 4]
```diff
@@ -45,7 +45,10 @@ export enum FeatureIdentifier {
   TERRAFORM_DESTROY = 'TERRAFORM_DESTROY',
   TERRAFORM_ROLLBACK = 'TERRAFORM_ROLLBACK',
   MAX_TOTAL_BUILDS = 'MAX_TOTAL_BUILDS',
+  MAX_TOTAL_SCANS = 'MAX_TOTAL_SCANS',
   MAX_BUILDS_PER_MONTH = 'MAX_BUILDS_PER_MONTH',
+  MAX_SCANS_PER_MONTH = 'MAX_SCANS_PER_MONTH',
+  TEST_INTELLIGENCE = 'TEST_INTELLIGENCE',
```
Comment (atmollo-harness) [inline src/framework/featureStore/FeatureIdentifier.ts:51]: Do we need this?

Comment (ahokinson) [reply src/framework/featureStore/FeatureIdentifier.ts:51]: Not sure why this is showing up in the diff since it appeared during a rebase. https://github.com/harness/harness-core-ui/blob/develop/src/modules/10-common/featureFlags.ts#L46

**Review by devinder-harness — APPROVED** [review]

**Review by devinder-harness — APPROVED** [review]

**Review by atmollo-harness — APPROVED** [review]

**Comment by ahokinson** [comment]
retrigger coverage

**Comment by ahokinson** [comment]
trigger cypress

**Comment by ahokinson** [comment]
trigger cypress

**Comment by ahokinson** [comment]
Coverage is totally reasonable and in line with existing module tests. There are a lot of conditional elements that create this "lack" of coverage. This should be force merged contingent on approval. Screen Shot 2022-10-25 at 15 22 49

**Review by atmollo-harness — DISMISSED** [review]

File: src/modules/70-pipeline/strings/strings.en.yaml [hunk 5]
```diff
@@ -1153,6 +1153,9 @@ featureRestriction:
   maxTotalBuilds90PercentLimit: You have used {{usagePercent}}% of builds included in the free plan. After 2,500 builds, the free plan will be restricted to 100 builds per month.
   maxBuildsPerMonth100PercentLimit: You have used 100 / 100 free builds this month. Consider upgrading for unlimited builds.
   numMonthlyBuilds: You have used {{count}} / {{limit}} free builds this month.
+  maxTotalScans90PercentLimit: You have used {{usagePercent}}% of scans included in the free plan. After 2,500 scans, the free plan will be restricted to 100 scans per month.
+  maxScansPerMonth100PercentLimit: You have used 100 / 100 free scans this month. Consider upgrading for unlimited builds.
```
Comment (mtran7) [inline src/modules/70-pipeline/strings/strings.en.yaml:1157]: do you mean upgrading for unlimited scans?

File: src/modules/33-auth-settings/pages/subscriptions/__tests__/STOUsageInfo.test.tsx [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    jest.spyOn(useUsage, 'useGetUsageAndLimit').mockReturnValue(data)
+
+    const { container } = render(
+      <TestWrapper>
+        <STOUsageInfo />
+      </TestWrapper>
+    )
+
+    expect(container).toMatchSnapshot()
```
[the remark this answers is not in the document]
Comment (ahokinson) [reply src/modules/33-auth-settings/pages/subscriptions/__tests__/STOUsageInfo.test.tsx:92]: Since they render differently, wouldn't this be the safer way to detect possible changes? If I explicitly check for something I'm mocking, it should always return true? This way we are ensuring the rendered html is as expected?

File: src/modules/70-pipeline/strings/strings.en.yaml (hunk 5 above)
Comment (ahokinson) [reply src/modules/70-pipeline/strings/strings.en.yaml:1157]: This is the standard language mirrored from CI. I have no doubt that it will change with further refinement, but the language will likely be determined by the documentation team. 

I think there is still some discussion around what the tiers are for STO, but currently this part isn't enforced and is more boiler plate.

**Review by mtran7 — DISMISSED** [review]

File: src/framework/featureStore/FeatureIdentifier.ts [hunk 7]
```diff
@@ -46,6 +46,9 @@ export enum FeatureIdentifier {
   TERRAFORM_ROLLBACK = 'TERRAFORM_ROLLBACK',
   MAX_TOTAL_BUILDS = 'MAX_TOTAL_BUILDS',
   MAX_BUILDS_PER_MONTH = 'MAX_BUILDS_PER_MONTH',
+  MAX_TOTAL_SCANS = 'MAX_TOTAL_SCANS',
+  MAX_SCANS_PER_MONTH = 'MAX_SCANS_PER_MONTH',
+  TEST_INTELLIGENCE = 'TEST_INTELLIGENCE',
```
Comment (chris-harness) [inline src/framework/featureStore/FeatureIdentifier.ts:51]: Where did the `TEST_INTELLIGENCE` come from? Seems unrelated

File: src/modules/10-common/hooks/useGetUsageAndLimit.ts [hunk 8]
```diff
@@ -286,6 +317,19 @@ function useGetUsage(module: ModuleName): UsageReturn {
...
+      case ModuleName.STO:
+        setUsageData({
+          usage: {
+            sto: {
+              activeDevelopers: stoUsageData?.data?.activeDevelopers,
+              activeScans: stoUsageData?.data?.activeScans
+            }
+          },
+          loadingUsage: loadingSTOUsage,
+          usageErrorMsg: ciUsageError?.message,
```
Comment (chris-harness) [inline src/modules/10-common/hooks/useGetUsageAndLimit.ts:329]:
```suggestion
          usageErrorMsg: stoUsageError?.message,
```

File: src/modules/10-common/hooks/useNavModuleInfo.ts [hunk 9]
```diff
@@ -49,7 +49,7 @@ interface ModuleInfo {
   icon: IconName
   label: StringKeys
   getHomePageUrl: (accountId: string) => string
-  featureFlagName: FeatureFlag
+  featureFlagName?: FeatureFlag
```
Comment (chris-harness) [inline src/modules/10-common/hooks/useNavModuleInfo.ts:52]: I'm not sure about this change. It could have widespread gotchas. Have you looked at all usages of it to verify it can be done safely?

File: src/modules/10-common/hooks/useNavModuleInfo.ts [hunk 10]
```diff
@@ -145,13 +144,20 @@ const useNavModuleInfo = (module: NavModuleName) => {
   const featureFlags = useFeatureFlags()
   const { licenseInformation } = useLicenseStore()
 
-  const { featureFlagName } = moduleInfoMap[module]
+  let shouldVisible: boolean
+  if (moduleInfoMap[module]?.featureFlagName != undefined) {
+    // eslint-disable-next-line
+    // @ts-ignore
```
Comment (chris-harness) [inline src/modules/10-common/hooks/useNavModuleInfo.ts:150]: We should not be `@ts-ignore`-ing anything

File: src/framework/featureStore/FeatureIdentifier.ts (hunk 7 above)
Comment (ahokinson) [reply src/framework/featureStore/FeatureIdentifier.ts:51]: It is unrelated. It keeps showing up in git changes despite not originating from my changes. I would imagine it appeared during a rebase. It's currently in develop, so I'm not sure why it's here.

File: src/modules/10-common/hooks/useNavModuleInfo.ts (hunk 9 above)
Comment (ahokinson) [reply src/modules/10-common/hooks/useNavModuleInfo.ts:52]: I agree. This was one of the changes that felt possibly like an overreach. But if the goal is not to use Feature Flags over time, this is a requirement given the existing implementation. It should probably be rewritten as more modules stop using Feature Flags to gate access, but it seemed like it would be a lot of effort just for STO.

Further down on https://github.com/harness/harness-core-ui/pull/11635/files/3741754c19b5a4ff128668d5c7e196b7baa0811b#diff-f6be1afd176cb1a2db952fb75097564719e42d219f213f1eb3cc40ecd3a2d933R148 you can see it gets handled identically to before **if** the feature flag still exists. So yes, the other usages are unaffected.

File: src/modules/10-common/hooks/useNavModuleInfo.ts (hunk 10 above)
Comment (ahokinson) [reply src/modules/10-common/hooks/useNavModuleInfo.ts:150]: For the life of me, I cannot get the linter to stop complaining about possible undefined variables even though it is **explicitly handled** by the if statement. I'd be glad to show you on my local branch. If this is removed, it will fail additional checks.

**Review by mtran7 — DISMISSED** [review]

File: src/modules/10-common/hooks/useNavModuleInfo.ts (hunk 10 above)
Comment (chris-harness) [reply src/modules/10-common/hooks/useNavModuleInfo.ts:150]: Yeah, this turned out to be a bit tricky to get TS to be happy. Here's what I came up with:
```
const shouldBeVisible = (module: NavModuleName, featureFlags: Partial<Record<FeatureFlag, boolean>>, licenseInformation: { [key: string]: ModuleLicenseDTO } | Record<string, undefined>) => {
  const featureFlagName = moduleInfoMap[module]?.featureFlagName
  return featureFlagName !== undefined ? !!featureFlags[featureFlagName] : licenseInformation[module]?.status === 'ACTIVE'
}

const useNavModuleInfo = (module: NavModuleName) => {
  const { accountId } = useParams<AccountPathProps>()
  const featureFlags = useFeatureFlags()
  const { licenseInformation } = useLicenseStore()

  return getModuleInfo(
    moduleInfoMap[module],
    accountId,
    licenseInformation[module]?.licenseType,
    shouldBeVisible(module, featureFlags, licenseInformation)
  ) as useNavModuleInfoReturnType
}
```
and reuse the `shouldBeVisible` function in `useNavModuleInfoMap`

You can even remove the existing `@ts-ignore` near the bottom of the file. It doesn't seem to be an issue anymore.

File: src/modules/45-dashboards/components/ModuleTagsFilter/__tests__/ModuleTagsFilter.test.tsx [hunk 11]
```diff
@@ -59,8 +64,7 @@ describe('ModuleTagsFilter', () => {
       CENG_ENABLED: true,
       CING_ENABLED: true,
       CDNG_ENABLED: false,
-      CFNG_ENABLED: false,
-      SECURITY: false
+      CFNG_ENABLED: false
     })
```
Comment (chris-harness) [inline src/modules/45-dashboards/components/ModuleTagsFilter/__tests__/ModuleTagsFilter.test.tsx:68]: You've altered what this test is doing, and it's doing a poorer job of testing code as a result. Instead, move `mockImport` inside the `test()`s and make it return inactive status for this test, active status for the previous test.

File: src/modules/45-dashboards/components/ModuleTagsFilter/__tests__/ModuleTagsFilter.test.tsx [hunk 12]
```diff
@@ -70,7 +74,7 @@ describe('ModuleTagsFilter', () => {
     expect(screen.getByText('buildsText')).toBeInTheDocument()
     expect(screen.queryByText('deploymentsText')).not.toBeInTheDocument()
     expect(screen.queryByText('common.purpose.cf.continuous')).not.toBeInTheDocument()
-    expect(screen.queryByText('common.purpose.sto.continuous')).not.toBeInTheDocument()
+    expect(screen.queryByText('common.purpose.sto.continuous')).toBeInTheDocument()
```
Comment (chris-harness) [inline src/modules/45-dashboards/components/ModuleTagsFilter/__tests__/ModuleTagsFilter.test.tsx:77]: You should be able to revert this with the previous comment's implementation

File: src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx [hunk 13]
```diff
@@ -52,8 +53,9 @@ export default function ExecutionTabs(props: ExecutionTabsProps): React.ReactEle
   const { view } = useQueryParams<ExecutionQueryParams>()
   const { updateQueryParams } = useUpdateQueryParams<ExecutionQueryParams>()
   const opaBasedGovernanceEnabled = useFeatureFlag(FeatureFlag.OPA_PIPELINE_GOVERNANCE)
-  const stoCDPipelineSecurityEnabled = useFeatureFlag(FeatureFlag.STO_CD_PIPELINE_SECURITY)
-  const stoCIPipelineSecurityEnabled = useFeatureFlag(FeatureFlag.STO_CI_PIPELINE_SECURITY)
+  const { licenseInformation } = useLicenseStore()
+  const stoCDPipelineSecurityEnabled = licenseInformation['STO']?.status === 'ACTIVE'
+  const stoCIPipelineSecurityEnabled = licenseInformation['STO']?.status === 'ACTIVE'
```
Comment (chris-harness) [inline src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx:58]: Combine these into one `const`

File: src/modules/72-templates-library/components/TemplateStudio/TemplateContext/TemplateContext.tsx [hunk 14]
```diff
@@ -1012,7 +1012,7 @@ export const TemplateProvider: React.FC<{
       isCIEnabled: licenseInformation['CI'] && CING_ENABLED,
       isCDEnabled: licenseInformation['CD'] && CDNG_ENABLED,
       isCFEnabled: licenseInformation['CF'] && CFNG_ENABLED,
-      isSTOEnabled: SECURITY_STAGE,
+      isSTOEnabled: licenseInformation['STO']?.status === 'ACTIVE',
```
Comment (chris-harness) [inline src/modules/72-templates-library/components/TemplateStudio/TemplateContext/TemplateContext.tsx:1015]: Why are other modules not checking for an active license here? Is there some subtle reason not to? Or is it just an existing bug?

File: src/modules/85-sto/RouteDestinations.tsx [hunk 15]
```diff
@@ -53,6 +68,117 @@ const moduleParams: ModulePathParams = {
...
+  licenseStateName: LICENSE_STATE_NAMES.STO_LICENSE_STATE,
+  startTrialRedirect: RedirectToModuleTrialHomeFactory(ModuleName.STO),
+  expiredTrialRedirect: RedirectToSubscriptionsFactory(ModuleName.STO)
+}
+
+featureFactory.registerFeaturesByModule('sto', {
+  features: [
+    FeatureIdentifier.MAX_TOTAL_BUILDS,
+    FeatureIdentifier.MAX_BUILDS_PER_MONTH,
+    FeatureIdentifier.ACTIVE_COMMITTERS
```
Comment (chris-harness) [inline src/modules/85-sto/RouteDestinations.tsx:83]: Seems like these should be different from CI

File: src/modules/85-sto/RouteDestinations.tsx [hunk 16]
```diff
@@ -53,6 +68,117 @@ const moduleParams: ModulePathParams = {
...
+      }
+    }
+
+    // If neither of limit breach/ warning/ overuse needs to be shown, return with an empty string.
+    // This will ensure no banner is shown
+    return {
+      message: () => '',
+      bannerType: BannerType.LEVEL_UP
+    }
+  }
```
Comment (chris-harness) [inline src/modules/85-sto/RouteDestinations.tsx:177]: There's an awful lot of reused logic in here. A reusable function would be nice to have, though not trivial to implement.

File: src/modules/85-sto/RouteDestinations.tsx [hunk 17]
```diff
@@ -136,9 +264,8 @@ const RedirectToProjectOverviewPage = (): React.ReactElement => {
 }
 
 const RouteDestinations: React.FC = () => {
-  const isV2 = useFeatureFlag(FeatureFlag.STO_API_V2)
   const { NG_SETTINGS } = useFeatureFlags()
-  const RemoteSTOApp = lazy(() => (isV2 ? import(`stoV2/App`) : import(`sto/App`)))
+  const RemoteSTOApp = lazy(() => import(`stoV2/App`))
 
   return (
     <>
```
Comment (chris-harness) [inline src/modules/85-sto/RouteDestinations.tsx:271]: There are several `licenseRedirectData` props that are still commented-out below

File: src/framework/featureStore/FeatureIdentifier.ts (hunk 7 above)
Comment (chris-harness) [reply src/framework/featureStore/FeatureIdentifier.ts:51]: I don't see it on `develop`. Might have been added then removed, or something? But as long as this is showing up, it's going to have unintended side effects when merged, so I'd get rid of it.

Comment (chris-harness) [reply src/framework/featureStore/FeatureIdentifier.ts:51]: https://github.com/harness/harness-core-ui/blob/develop/src/framework/featureStore/FeatureIdentifier.ts

File: src/modules/10-common/hooks/useNavModuleInfo.ts (hunk 9 above)
Comment (chris-harness) [reply src/modules/10-common/hooks/useNavModuleInfo.ts:52]: It looks like it's kept internal to that one file, so yeah, seems ok 👍🏻

File: src/modules/72-templates-library/components/TemplateStudio/TemplateContext/TemplateContext.tsx (hunk 14 above)
Comment (ahokinson) [reply src/modules/72-templates-library/components/TemplateStudio/TemplateContext/TemplateContext.tsx:1015]: It simply hasn't been implemented. We are the first to be checking for active licenses. At the moment, licenses don't really gate any behavior and SEs are required to add licenses AND feature flags to new customers to get functionality working.

File: src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx [hunk 18]
```diff
@@ -232,7 +231,7 @@ export default function ExecutionTabs(props: ExecutionTabsProps): React.ReactEle
     })
   }
 
-  if ((isCD && stoCDPipelineSecurityEnabled) || ((isCI || isSTO) && stoCIPipelineSecurityEnabled)) {
+  if ((isCD || isCI || isSTO) && stoPipelineSecurityEnabled) {
```
Comment (chris-harness) [inline src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx:234]: Yes, much clearer logic! 🎉

**Review by chris-harness — APPROVED** [review]

**Review by atmollo-harness — APPROVED** [review]

**Review by mtran7 — DISMISSED** [review]

**Review by tan-nhu — APPROVED** [review]

**Review by pranil-harness — DISMISSED** [review]

**Comment by ahokinson** [comment]
trigger cypress

**Comment by ahokinson** [comment]
retrigger jest

File: src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx (hunk 18 above)
Comment (atmollo-harness) [reply src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx:234]: Shouldn't we remove (isCD || isCI || isSTO) and just rely on the license being enabled? The new top level projects -> executions view isn't technically in any module https://qa.harness.io/ng/#/account/BdsgiWzwT7CQFeJl9XkQ3A/home/orgs/default/projects/STO/deployments
So it's technically not in CI CD or STO so this tab won't show up

Comment (ahokinson) [reply src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx:234]: I would prefer not to make that change here. It's still following the pattern CI has set for other inclusions in the tab list.

Comment (chris-harness) [reply src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx:234]: I agree. Let's limit scope creep so we can get this in with minimal risk. We have a separate ticket we can address that in.

Comment (atmollo-harness) [reply src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx:234]: Okay, alternatively this can be changed to `||` rather than `&&` but If we have the other ticket to track it it should be fine. Im concerned that someone is going to hit that feature flag in prod and make the Security Tests tab disappear in the new view.

**Review by atmollo-harness — DISMISSED** [review]

**Review by chris-harness — APPROVED** [review]

File: src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx (hunk 18 above)
Comment (chris-harness) [reply src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx:234]: @atmollo-harness the feature flags are all being removed in this PR, so that can't happen.

Comment (atmollo-harness) [reply src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx:234]: @chris-harness `const stoPipelineSecurityEnabled = licenseInformation['STO']?.status === 'ACTIVE'` yea the FF are being removed and the new check is going to be seeing if there is an active license

Comment (ahokinson) [reply src/modules/70-pipeline/pages/execution/ExecutionLandingPage/ExecutionTabs/ExecutionTabs.tsx:234]: Sales engineers are aware of the change. We know all the accounts that are switching over.
